### ISSUE TITLE
[suggestion] Could some functions be synchronized between MultiLayerNetwork and ComputationGraph?

### ISSUE THREAD

**Issue by yquemener** [deeplearning4j/deeplearning4j#5761]
Some functions do almost the same thing across these two classes of models but have slightly different syntax and/or arguments, return types. Maybe the following functions could be made abstract and placed inside Model?

- feedForwardToLayer vs feedForward(..., tillLayerIndex)
- fit(INDArray, INDArray) vs fit(INDArray[], INDArray[])
- output(...) that returns either INDArray[] or INDArray (maybe this one makes more sense)

**Comment by AlexDBlack** [comment]
Related (but not duplicate): https://github.com/deeplearning4j/deeplearning4j/issues/5397

I'd be hesitant to put them on Model (without other changes first) as currently we have ```Layer extends Model``` which means we'd be making even more no-op methods on the Layers.

**Comment by yquemener** [comment]
Yes, I should have looked at all the things that are inherited from Model before making that proposal. Please do not hesitate to dismiss it as impractical, I don't know the whole history of DL4J or the reason why Layer is a Model. 

Seeing how I often repeat code like `if(model instanceof ComputationGraph) in Discover, I am tempted to make a kind of wrapper for ComputationGraph, MultiLayerNetwork but also ParagraphVector, as things like word vectors do not inherit Model but have a similar train/infer logic.

**Comment by AlexDBlack** [comment]
The idea has been discussed a number of times in the past, and I think it's a worthwhile goal.
However, we would need to think carefully about how to implement it (perhaps this is best implemented along with some larger-scale refactoring)... the Layer API has more than enough methods already, and I'd like to simplify it further - not add more methods if possible :)

**Comment by saudet** [comment]
Maybe we could have a BaseModel that throws UnsupportedOperationException for anything the layers don't support and make AbstractLayer and others extend that? It would still add a lot of noise the API though..

**Comment by yquemener** [comment]
I was thinking of making something like a Trainable interface in Discover, so that I can integrate Word2Vec and Paragraph2Vec with it too.

**Comment by cpoerschke** [comment]
> ... often repeat code like `if(model instanceof ComputationGraph) in Discover, I am tempted to make a kind of wrapper for ComputationGraph, MultiLayerNetwork but also ParagraphVector ...

In #5428 I'd been pointed in the direction of [NetworkUtils](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/util/NetworkUtils.java) but looks like that would be suitable only for any ComputationGraph+MultiLayerNetwork portion since [ParagraphVectors](https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/java/org/deeplearning4j/models/paragraphvectors/ParagraphVectors.java) lives elsewhere.
